contracts/extensions/TokenSupplier.sol

```diff
@@ -0,0 +1,188 @@
+
+  /// @notice Called when uninstalling the extension
+  function uninstall() public override auth {
+    selfdestruct(address(uint160(address(colony))));
+  }
+
+  /// @notice Initialise the extension, must be called before any tokens can be issued
+  /// @param _tokenSupplyCeiling Total amount of tokens to issue
+  /// @param _tokenIssuanceRate Number of tokens to issue per day
+  function initialise(uint256 _tokenSupplyCeiling, uint256 _tokenIssuanceRate) public onlyRoot {
+    require(lastPinged == 0, "token-supplier-already-initialised");
+
+    tokenSupplyCeiling = _tokenSupplyCeiling;
+    tokenIssuanceRate = _tokenIssuanceRate;
+    lastPinged = block.timestamp;
+    lastRateUpdate = block.timestamp;
+  }
+
+  /// @notice Update the tokenSupplyCeiling, cannot set below current tokenSupply
+  /// @param _tokenSupplyCeiling Total amount of tokens to issue
+  function setTokenSupplyCeiling(uint256 _tokenSupplyCeiling) public initialised onlyRoot onlyContract {
+    try ColonyExtension(msg.sender).identifier() returns (bytes32 extensionId) {
+      IColonyNetwork colonyNetwork = IColonyNetwork(colony.getColonyNetwork());
+      address installation = colonyNetwork.getExtensionInstallation(extensionId, address(colony));
+      require(installation == msg.sender, "token-supplier-not-managed-extension");
+      require(extensionId == VOTING_HYBRID, "token-supplier-cannot-set-value");
+    } catch {
+      require(false, "token-supplier-no-identifier");
+    }
+
+    tokenSupplyCeiling = _tokenSupplyCeiling;
+
+    emit TokenSupplyCeilingSet(tokenSupplyCeiling);
+  }
+
+  /// @notice Update the tokenIssuanceRate
+  /// @param _tokenIssuanceRate Number of tokens to issue per day
+  function setTokenIssuanceRate(uint256 _tokenIssuanceRate) public initialised onlyRoot onlyContract {
+    try ColonyExtension(msg.sender).identifier() returns (bytes32 extensionId) {
+      IColonyNetwork colonyNetwork = IColonyNetwork(colony.getColonyNetwork());
+      address installation = colonyNetwork.getExtensionInstallation(extensionId, address(colony));
+      require(installation == msg.sender, "token-supplier-not-managed-extension");
+      require(
+        extensionId == VOTING_HYBRID || (
+          extensionId == VOTING_REPUTATION &&
+          block.timestamp - lastRateUpdate >= 4 weeks &&
+          _tokenIssuanceRate <= add(tokenIssuanceRate, tokenIssuanceRate / 10) &&
+          _tokenIssuanceRate >= sub(tokenIssuanceRate, tokenIssuanceRate / 10)
+        ),
+        "token-supplier-cannot-set-value"
+      );
+    } catch {
+      require(false, "token-supplier-no-identifier");
+    }
+
+    // Issue any outstanding tokens under the previous rate
+    issueTokens();
+
+    tokenIssuanceRate = _tokenIssuanceRate;
+    lastRateUpdate = block.timestamp;
+
+    emit TokenIssuanceRateSet(tokenIssuanceRate);
+  }
+
+  /// @notice Issue the appropriate amount of tokens
+  function issueTokens() public initialised {
```

Comment:
This function previously had a check in it that threw if there were no tokens to issue. By removing it, even if no tokens are issued, `lastPinged` is updated, which can be used to prevent any tokens being minted at all by repeatedly calling this function if the issuance rate is sufficiently low.

Why was the check removed?

Reply:
Oh good point. I removed it because it had a bad interaction with `setTokenIssuanceRate` -- if you updated the issuance rate when there were no tokens to issue, that call would fail. In addition, I have a mild preference for these types of functions just "passing through" if there is nothing to issue, vs throwing an error (0 tokens is a valid number to issue, it doesn't really represent a failure condition). I think what I'll do is not update the issuance rate if there are no tokens to issue, to address your valid concern.